CurrencyConverterControl/Data/CurrencyDataProvider.cs

```diff
@@ -0,0 +1,137 @@
+
+                    string endpoint = String.Format("{0}convert?access_key={1}&from={2}&to={3}&amount={4}", baseURl, access_key, currencyfrom, currencyto, amount);
+                    HttpResponseMessage response = await httpClient.GetAsync(endpoint).ConfigureAwait(false);
+
+                    if (response.IsSuccessStatusCode)
+                    {
+                        string content = await response.Content.ReadAsStringAsync();
+                        JObject json = JObject.Parse(content);
+                        double value = (double)json["result"];                        
+                        return value;
+                    }
+                    else
+                    {
+                        Console.WriteLine("Request failed with status code: " + response.StatusCode);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Error: {ex.Message}");
+            }
+            return -1;
+        }
+        /// <summary>
+        /// This method is used to get List of currencies which holds Currency Code and Flag 
+        /// </summary>
+        /// <returns>Returns a List of Currency Objects </returns>
+        /// <exception cref="DirectoryNotFoundException">When Path of the file is not found</exception>
+        public List<Currency> GetCurrenciesData()
+        {
+            string filePath = @"CurrencyListXML.xml";
+
+            // Call the method to read data
+            return DeserializeFromXml(filePath);
+        }
+        private List<Currency> DeserializeFromXml(string inputPath)
+            {
+               CurrencyList currencies = new CurrencyList();
+
+                XmlSerializer serializer = new XmlSerializer(typeof(CurrencyList));
+
+                try
+                {
+                    using (TextReader reader = new StreamReader(inputPath))
+                    {
+                        currencies = (CurrencyList)serializer.Deserialize(reader);
+                    }
+
+                    Console.WriteLine($"XML file successfully deserialized from: {inputPath}");
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"Error deserializing XML file: {ex.Message}");
+                }
+
+                return currencies.CurrenciesList;
+            }
+
+        /// <summary>
+        /// Gets the list of convertion Rates for two currencies
+        /// </summary>
+        /// <param name="currencyfrom"></param>
+        /// <param name="currencyto"></param>        
+        /// <returns>Returns the value of the convertion at that moment</returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        /// <exception cref="System.NullReferenceException">Happens when your monthly usage limit has been reached.  </exception>
+        public async Task<double> GetConversionRate(string currencyfrom, string currencyto)
```

Comment:
@techNIKtos  i have thought of that solution, but the fact that the rates might change stepped me back.  I will implement it as a solution. 
thanks

Reply:
@rexhinahoxha , yes, they can change. It's good that you thought about it.

But there is no point in doing this with every action. In addition, this approach has low performance. In this case, you can check how much time has passed since the last rates update and if it exceeds, say, a minute, then request the data again.